### ISSUE TITLE
Adding progress reporting to Controllers (components) [CARROT-162]

### ISSUE THREAD

**Issue by dweiss** [carrot2/carrot2#368]
Processing of a query (especially when documents are extracted from online system e.g. Yahoo!, MSN) can last a while. It would be really practical to have a way to track progress of processing. Perhaps some ProgressListener and pair of add/removeProgressListener in Controllers (maybe it should be implemented even in Component level, can't really tell now).

This can be later used in all GUI apps to show progress bar etc.

---
Issue: CARROT-162 (migrated from JIRA), created by Urszula Krukar, resolved May 01 2008
Linked issues:
 - CARROT-187

**Comment by dweiss** [comment]
Comment by Stanisław Osiński (@stanislawosinski) (migrated from JIRA)

Very good point! I guess there are two types of progress: a) finishing processing by each component in the pipeline (tracked by the controller), b) progress of processing within individual components (tracked within components). Both are worth supporting in the API, though in practice only progress for option a) will be always available – some components may not be able to provide more detailed progress information.

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

Good point, but very hard to achieve in practice. Progress between components is very coarse-level, so it will be barely visible, progress at component level is not practical – for example querying a search engine takes a long pause (concurrent queries) and a series of rapid updates (results). 

I don't see how this could be made functional.

**Comment by dweiss** [comment]
Comment by Stanisław Osiński (@stanislawosinski) (migrated from JIRA)

Adding a simple interface handling both cases should not be too much of a problem, that's for sure. If we were to visualize the progress between components as a progress bar, that would look strange indeed because most of the time we'd have 2 components. But if we were to show in the status line something along the lines: "Collecting documents from &lt;<YahooAPI&gt;>" or "Clustering with &lt;<STC&gt;>", that would show some kind of progress.

When it comes to intra-component progress, as Dawid said, some components will not provide it at all. But I can imagine that e.g. if you're fetching documents from a Lucene index sequentially, you could update the progress every 10% of docs or something. But that's again a bit theoretical example (Lucene is quite fast). On the other hand, for certain clustering algorithms you could provide some more fine-grained progress related to different stages of the clustering like tokenization, factorization etc (Lingo and Lingo3G have many of them). This would be hard to express in terms of % of progress, but the total number of stages will be know upfront.

To sum up, this kind of feature is certainly not a M1 or M2 feature, but if there's nothing to do in M3, I'm happy to implement support for it :-)

**Comment by dweiss** [comment]
Comment by Urszula Krukar (migrated from JIRA)

If it's really that complicated (and the progress flow won't be fluent at all) then all I need is this status line feature. Right now, without any progress indication, it looks like this: 'Progress' clicked, 2-3 seconds waiting for editor to show up, editors shows up. And the algorithm is really simple. I can imaging more complicated algorithms taking longer. Any indication, that "something's happening" would be good. 

I can always use indeterminate progress bar for this (not sure if Eclipse has anything like this, will check later, when this is an issue).

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

It's much like with Internet Explorer – something is moving, but it's not actually doing anything (it's time-based).... :)

**Comment by dweiss** [comment]
Comment by Stanisław Osiński (@stanislawosinski) (migrated from JIRA)

This is how it's actually implemented in the current (Swing) browser too. I think some feedback that the processing has started is important, even if then it looks like in IE.
